python/ql/src/experimental/semmle/python/security/injection/CsvInjection.qll

```diff
@@ -10,11 +11,21 @@ import semmle.python.dataflow.new.RemoteFlowSources
 class CsvInjectionFlowConfig extends TaintTracking::Configuration {
   CsvInjectionFlowConfig() { this = "CsvInjectionFlowConfig" }
 
-  override predicate isSource(DataFlow::Node source) {
-    source instanceof RemoteFlowSource
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) { sink = any(CsvWriter cw).getAnInput() }
+
+  override predicate isSanitizerGuard(DataFlow::BarrierGuard guard) {
+    guard instanceof StartsWithCheck or
+    guard instanceof StringConstCompare
   }
+}
+
+class StartsWithCheck extends DataFlow::BarrierGuard {
+  StartsWithCheck() { this.(CallNode).getNode().getFunc().(Attribute).getName() = "startswith" }
 
-  override predicate isSink(DataFlow::Node sink) {
-    exists(CsvWriter csvwriter | sink = csvwriter.getAnInput())
+  override predicate checks(ControlFlowNode node, boolean branch) {
+    node = this.(CallNode).getNode().getFunc().(Attribute).getObject().getAFlowNode() and
+    branch = true
   }
 }
```

Comment:
If you make `StartsWithCheck` also extend `DataFlow::MethodCallNode`, you can simplify both the charpred and the `checks` predicate.

Reply:
Need to remove the `isSink` predicate?

Reply:
No, that is just how the diff shows. I am only talking about the class `StartsWithCheck`.

Reply:
I modified it.